pkg/cgroups/manager/manager.go

```diff
@@ -0,0 +1,392 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package manager
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cilium/cilium/pkg/cgroups"
+	v1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	nodetypes "github.com/cilium/cilium/pkg/node/types"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "cgroup-manager")
+	// example cgroup path in kubernetes environments
+	// /run/cilium/cgroupv2/kubepods/burstable/pod1858680e-b044-4fd5-9dd4-f137e30e2180/e275d1a37782ab30008aa3ae6666cccefe53b3a14a2ab5a8dc459939107c8c0
+	defaultCgroupBasePath = cgroups.GetCgroupRoot() + "/kubepods"
+	// example cgroup path in nested kubernetes environments like kind
+	// /run/cilium/cgroupv2/kubelet/kubepods/pod4841248b-fc2f-41f4-9981-a685bf840ab5/d8f227cc24940cfdce8d8e601f3b92242ac9661b0e83f0ea57fdea1cb6bc93ec
+	defaultNestedCgroupBasePath = cgroups.GetCgroupRoot() + "/kubelet" + "/kubepods"
+	cgroupBasePaths             = []string{defaultCgroupBasePath, defaultNestedCgroupBasePath}
+	// Channel buffer size for pod events in order to not block callers
+	podEventsChannelSize = 20
+)
+
+// Pod events processed by CgroupManager
+const (
+	podAddEvent = iota
+	podUpdateEvent
+	podDeleteEvent
+	podGetMetadataEvent
+)
+
+// CgroupManager maintains Kubernetes and low-level metadata (cgroup path and
+// cgroup id) for local pods and their containers. In order to do that, it defines
+// and implements callback functions that are called on Kubernetes pod watcher events.
+// It also exposes APIs to read the saved metadata.
+//
+// The manager's internals are synchronized via a channel, and must not be
+// accessed/updated outside this channel.
+//
+// During initialization, the manager checks for a valid base cgroup path from
+// known defaults. In case of environments using non-default paths, manager will
+// fail to get a valid cgroup base path, and ignore all the subsequent pod events.
+type CgroupManager struct {
+	// Map of pod metadata indexed by their UIDs
+	podMetadataById map[podUID]*podMetadata
```

Comment:
I feel like we are going to want a `cilium cgroups list` CLI command to introspect this, if there's ever trouble while trying to map the socket messages to pods/containers. Did you ever find yourself needing something like this during development?

Might be an interesting followup.

Reply:
Good feedback! Having a CLI command to inspect the internals sounds handy for debugging. Added to the follow-ups.